arwen/host/execution.go

```diff
@@ -189,33 +177,108 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
 }
 
 func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (*vmcommon.VMOutput, error) {
-	host.PushState()
-	defer host.PopState()
+	bigInt := host.BigInt()
+	output := host.Output()
+	runtime := host.Runtime()
+	storage := host.Storage()
 
-	host.InitState()
+	bigInt.PushState()
+	bigInt.InitState()
+
+	output.PushState()
+	output.CensorVMOutput()
+
+	runtime.PushState()
+	runtime.InitStateFromContractCallInput(input)
 
-	host.Runtime().InitStateFromContractCallInput(input)
-	host.Storage().SetAddress(host.Runtime().GetSCAddress())
+	storage.PushState()
+	storage.SetAddress(host.Runtime().GetSCAddress())
 
-	err := host.execute(input)
+	// Perform a value transfer to the called SC. If the execution fails, this
+	// transfer will not persist.
+	err := output.Transfer(input.RecipientAddr, input.CallerAddr, 0, input.CallValue, nil)
 	if err != nil {
-		return nil, err
+		// Execution failed: restore contexts as if the execution didn't happen,
+		// but first create a vmOutput to capture the error.
+		vmOutput := output.CreateVMOutputInCaseOfError(err)
+
+		bigInt.PopSetActiveState()
+		output.PopSetActiveState()
+		runtime.PopSetActiveState()
+
+		return vmOutput, err
+	}
+
+	err = host.execute(input)
+	if err != nil {
+		// Execution failed: restore contexts as if the execution didn't happen,
+		// but first create a vmOutput to capture the error.
+		vmOutput := output.CreateVMOutputInCaseOfError(err)
+
+		bigInt.PopSetActiveState()
+		output.PopSetActiveState()
+		runtime.PopSetActiveState()
+
+		return vmOutput, err
 	}
 
+	// Extract the VMOutput produced by the execution in isolation, before
+	// restoring the contexts. This needs to be done before popping any state
+	// stacks.
 	vmOutput := host.Output().GetVMOutput()
 
+	// Execution successful: restore the previous context states, except Output,
+	// which will merge the current state (VMOutput) with the initial state.
+	bigInt.PopSetActiveState()
+	output.PopMergeActiveState()
+	runtime.PopSetActiveState()
+	storage.PopSetActiveState()
+
 	return vmOutput, nil
 }
 
 func (host *vmHost) ExecuteOnSameContext(input *vmcommon.ContractCallInput) error {
+	bigInt := host.BigInt()
+	output := host.Output()
 	runtime := host.Runtime()
+
+	// Back up the states of the contexts (except Storage, which isn't affected
+	// by ExecuteOnSameContext())
+	bigInt.PushState()
+	output.PushState()
 	runtime.PushState()
-	defer runtime.PopState()
 
 	runtime.InitStateFromContractCallInput(input)
-	err := host.execute(input)
 
-	return err
+	// Perform a value transfer to the called SC. If the execution fails, this
+	// transfer will not persist.
+	err := output.Transfer(input.RecipientAddr, input.CallerAddr, 0, input.CallValue, nil)
+	if err != nil {
+		// Execution failed: restore contexts as if the execution didn't happen.
+		bigInt.PopSetActiveState()
```

Comment:
same as above. duplicated code plus use defer if err ->

Reply:
Rewrote this using `defer` and handled in a separate function, depending on `err`.